pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerCornerCaseTest.java

```diff
@@ -0,0 +1,356 @@
+        producer.close();
+        msgBuilder.value("msg-1").sendAsync().exceptionally(ex -> {
+            log.warn("expected error", ex);
+            return null;
+        }).join();
+
+        // Verify: message payload has been released.
+        Awaitility.await().untilAsserted(() -> {
+            assertEquals(producer.getPendingQueueSize(), 0);
+        });
+        assertEquals(msgBuilder.payload.refCnt(), 1);
+
+        // cleanup.
+        msgBuilder.release();
+        assertEquals(msgBuilder.payload.refCnt(), 0);
+        admin.topics().delete(topicName);
+        //admin.topics().deleteAsync(topicName).get(10, TimeUnit.SECONDS); TODO fix bug.
+    }
+
+    @DataProvider
+    public Object[][] failedInterceptAt() {
+        return new Object[][]{
+          {"eligible"},
+          {"beforeSend"},
+          {"onSendAcknowledgement"},
+        };
+    }
+
+    @Test(dataProvider = "failedInterceptAt")
+    public void testInterceptorError(String method) throws Exception {
+        final String topicName = BrokerTestUtil.newUniqueName("persistent://public/default/tp_");
+        admin.topics().createNonPartitionedTopic(topicName);
+        ProducerImpl<String> producer = (ProducerImpl<String>) pulsarClient.newProducer(Schema.STRING).topic(topicName)
+                .intercept(
+
+                new ProducerInterceptor() {
+                    @Override
+                    public void close() {
+                        if (method.equals("close")) {
+                            throw new RuntimeException("Mocked error");
+                        }
+                    }
+
+                    @Override
+                    public boolean eligible(Message message) {
+                        if (method.equals("eligible")) {
+                            throw new RuntimeException("Mocked error");
+                        }
+                        return false;
+                    }
+
+                    @Override
+                    public Message beforeSend(Producer producer, Message message) {
+                        if (method.equals("beforeSend")) {
+                            throw new RuntimeException("Mocked error");
+                        }
+                        return message;
+                    }
+
+                    @Override
+                    public void onSendAcknowledgement(Producer producer, Message message, MessageId msgId,
+                                                      Throwable exception) {
+                        if (method.equals("onSendAcknowledgement")) {
+                            throw new RuntimeException("Mocked error");
+                        }
+
+                    }
+                }).create();
+
+        MsgPayloadTouchableMessageBuilder<String> msgBuilder = newMessage(producer);
+        try {
+            msgBuilder.value("msg-1").sendAsync().get(3, TimeUnit.SECONDS);
+        } catch (Exception ex) {
+            log.warn("Intercept error", ex);
```

Comment:
Yet another unnecessary exception stack logs, which is noisy. And you should call `fail()` in the try block.

Reply:
Solved